src/server/pfs/server/master.go

```diff
@@ -139,6 +157,119 @@ func (d *driver) finishCommits(ctx context.Context) (retErr error) {
 	)
 }
 
+func (d *driver) manageBranches(ctx context.Context, repoKey string) error {
+	ctx, cancel := pctx.WithCancel(ctx)
+	defer cancel()
+	type cronTrigger struct {
+		cancel context.CancelFunc
+		spec   string
+	}
+	cronTriggers := make(map[string]*cronTrigger)
+	defer func() {
+		for _, ct := range cronTriggers {
+			ct.cancel()
+		}
+	}()
+	return d.branches.ReadOnly(ctx).WatchByIndexF(pfsdb.BranchesRepoIndex, repoKey, func(ev *watch.Event) error {
+		if ev.Type == watch.EventError {
+			return ev.Err
+		}
+		key := string(ev.Key)
+		if ev.Type == watch.EventDelete {
+			if ct, ok := cronTriggers[key]; ok {
+				ct.cancel()
+			}
+			delete(cronTriggers, key)
+			return nil
+		}
+		branchInfo, err := d.inspectBranch(ctx, pfsdb.ParseBranch(key))
+		if err != nil {
+			return err
+		}
+		if branchInfo.Trigger == nil || branchInfo.Trigger.CronSpec == "" {
+			return nil
+		}
+		// Only create a new goroutine if one doesn't already exist or the spec changed.
+		if ct, ok := cronTriggers[key]; ok {
+			if ct.spec == branchInfo.Trigger.CronSpec {
+				return nil
+			}
+			ct.cancel()
+			delete(cronTriggers, key)
+		}
+		ctx, cancel := pctx.WithCancel(ctx)
+		cronTriggers[key] = &cronTrigger{
+			cancel: cancel,
+			spec:   branchInfo.Trigger.CronSpec,
+		}
+		go func() {
+			backoff.RetryUntilCancel(ctx, func() error { //nolint:errcheck
+				return d.runCronTrigger(ctx, branchInfo.Branch)
+			}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+				log.Error(ctx, "error running cron trigger", zap.String("branch", key), zap.Error(err), zap.Duration("retryAfter", d))
+				return nil
+			})
+		}()
+		return nil
+	})
+}
+
+func (d *driver) runCronTrigger(ctx context.Context, branch *pfs.Branch) error {
+	branchInfo, err := d.inspectBranch(ctx, branch)
+	if err != nil {
+		if pfsserver.IsBranchNotFoundErr(err) {
+			log.Info(ctx, "branch no longer exists; exiting cron trigger", zap.String("branch", pfsdb.BranchKey(branch)))
+			return nil
+		}
+		return err
+	}
+	schedule, err := cronutil.ParseCronExpression(branchInfo.Trigger.CronSpec)
+	if err != nil {
+		return err
+	}
+	// Use the current head commit start time as the previous tick.
+	// This prevents the timer from restarting if the master restarts.
+	ci, err := d.inspectCommit(ctx, branchInfo.Head, pfs.CommitState_STARTED)
```

Comment:
What happens on an empty branch?

Reply:
Also should we take the minimum of this and the current time?

Reply:
good point on the empty branch edge case - the semantics of triggering a branch -> nil doesn't make sense to me. We probably should exit if `branchInfo.Head == nil`

Reply:
Branch heads cannot be nil since they always start with an empty commit as head. I'm not sure I understand the minimum suggestion. Is the concern clock drift? If not, then the commit start time will always be less than the current time.